onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc

```diff
@@ -31,18 +33,17 @@ namespace onnxruntime {
     }                                                           \
   } while (0)
 
-TensorrtExecutionProvider::TensorrtExecutionProvider()
-    : IExecutionProvider{onnxruntime::kTensorrtExecutionProvider} {
-  DeviceAllocatorRegistrationInfo trt_device_info({OrtMemTypeCPU, [](int) {
-                                                     return std::make_unique<TensorrtPinnedAllocator>();
-                                                   },
-                                                   std::numeric_limits<size_t>::max()});
-  InsertAllocator(CreateAllocator(trt_device_info));
-  DeviceAllocatorRegistrationInfo default_device_info({OrtMemTypeDefault, [](int) {
-                                                         return std::make_unique<TensorrtAllocator>();
-                                                       },
-                                                       std::numeric_limits<size_t>::max()});
-  InsertAllocator(CreateAllocator(default_device_info));
+TensorrtExecutionProvider::TensorrtExecutionProvider(const TensorrtExecutionProviderInfo& info)
+    : IExecutionProvider{onnxruntime::kTensorrtExecutionProvider}, device_id_(info.device_id) {
+  CUDA_CALL_THROW(cudaSetDevice(device_id_));
+
+  DeviceAllocatorRegistrationInfo default_allocator_info(
+      {OrtMemTypeDefault, [](int id) { return std::make_unique<TensorrtAllocator>(id); }, std::numeric_limits<size_t>::max()});
+  InsertAllocator(CreateAllocator(default_allocator_info, device_id_));
```

Comment:
is it possible to reuse the allocators defined in cuda_allocator?

Reply:
TensorRT EP has higher priority than CUDA, so it needs to register its allocator first. Ke is doing more allocator factoring. After that TensorRT and CUDA may use same GPU allocator.

Reply:
I mean remove the tensorrt_allocator.h/.cc, 
#include "cuda_allocator.h"
copy same code from cuda EP:
  DeviceAllocatorRegistrationInfo default_allocator_info(
      {OrtMemTypeDefault, [](int id) { return std::make_unique<CUDAAllocator>(id); }, std::numeric_limits<size_t>::max()});
  InsertAllocator(CreateAllocator(default_allocator_info, device_id_));

---
In reply to: [311861343](https://github.com/microsoft/onnxruntime/pull/1561#discussion_r311861343) [](ancestors = 311861343)

Reply:
in that case ORT will show OrtAllocatorInfo: allocator already registered since TRT registered the allocator first then CUDA is trying to register the same allocator again later